libr/crypto/p/crypto_bech32.c

```diff
@@ -0,0 +1,217 @@
+		i++;
+	}
+	if (i + 7 + data_len > 90) {
+		return 0;
+	}
+	chk = bech32_polymod_step (chk);
+	while (*hrp != 0) {
+		chk = bech32_polymod_step (chk) ^ (*hrp & 0x1f);
+		*(output++) = *(hrp++);
+	}
+	*(output++) = '1';
+	for (i = 0; i < data_len; i++) {
+		if (*data >> 5) {
+			return 0;
+		}
+		chk = bech32_polymod_step (chk) ^ (*data);
+		*(output++) = charset[*(data++)];
+	}
+	for (i = 0; i < 6; i++) {
+		chk = bech32_polymod_step (chk);
+	}
+	chk ^= bech32_final_constant (enc);
+	for (i = 0; i < 6; i++) {
+		*(output++) = charset[(chk >> ((5 - i) * 5)) & 0x1f];
+	}
+	*output = 0;
+	return 1;
+}
+
+static bech32_encoding bech32_decode(char *hrp, uint8_t *data, int data_len, const char *input) {
+	uint32_t chk = 1;
+	size_t i, input_len = strlen (input);
+	int have_lower = 0, have_upper = 0;
+	if (input_len < 8 || input_len > 90) {
+		return BECH32_ENCODING_NONE;
+	}
+	data_len = 0;
+	while (data_len < input_len && input[(input_len - 1) - data_len] != '1') {
+		(data_len)++;
+	}
+	size_t hrp_len = input_len - (1 + data_len);
+	if (1 + data_len >= input_len || data_len < 6) {
+		return BECH32_ENCODING_NONE;
+	}
+	data_len -= 6;
+	for (i = 0; i < hrp_len; i++) {
+		int ch = input[i];
+		if (ch < 33 || ch > 126) {
+			return BECH32_ENCODING_NONE;
+		}
+		if (ch >= 'a' && ch <= 'z') {
+			have_lower = 1;
+		} else if (ch >= 'A' && ch <= 'Z') {
+			have_upper = 1;
+			ch = (ch - 'A') + 'a';
+		}
+		hrp[i] = ch;
+		chk = bech32_polymod_step (chk) ^ (ch >> 5);
+	}
+	hrp[i] = 0;
+	chk = bech32_polymod_step (chk);
+	for (i = 0; i < hrp_len; i++) {
+		chk = bech32_polymod_step (chk) ^ (input[i] & 0x1f);
+	}
+	i++;
+	while (i < input_len) {
+		int v = (input[i] & 0x80)? -1: charset_rev[(int)input[i]];
+		if (input[i] >= 'a' && input[i] <= 'z')
+			have_lower = 1;
+		if (input[i] >= 'A' && input[i] <= 'Z')
+			have_upper = 1;
+		if (v == -1) {
+			return BECH32_ENCODING_NONE;
+		}
+		chk = bech32_polymod_step (chk) ^ v;
+		if (i + 6 < input_len) {
+			data[i - (1 + hrp_len)] = v;
+		}
+		i++;
+	}
+	if (have_lower && have_upper) {
+		return BECH32_ENCODING_NONE;
+	}
+	if (chk == bech32_final_constant (BECH32_ENCODING_BECH32)) {
+		return BECH32_ENCODING_BECH32; // wtf?
+		if (chk == bech32_final_constant (BECH32_ENCODING_BECH32M)) {
+			return BECH32_ENCODING_BECH32M;
+		}
+		return BECH32_ENCODING_NONE;
+	}
+	return BECH32_ENCODING_NONE;
+}
+
+static bool bech32_set_key(RCryptoJob *cj, const ut8 *key, int keylen, int mode, int direction) {
+	cj->key_len = keylen;
+	memcpy (cj->key, key, keylen);
+	cj->dir = direction;
+	return true;
+}
+
+static int bech32_get_key_size(RCryptoJob *cj) {
+	return cj->key_len;
+}
+
+static bool bech32_check(const char *algo) {
+	return !strcmp (algo, "bech32");
+}
+
+static bool update(RCryptoJob *cj, const ut8 *buf, int len) {
+	const int enc = BECH32_ENCODING_BECH32;
+	int hrp_size = hrplength (buf, len);
+	char *hrp = malloc (cj->key_len + 1); // HRP need to be null-terminated
```

Comment:
check if hrp is null or not
```suggestion
	char *hrp = malloc (cj->key_len + 1); // HRP need to be null-terminated
	if (!hrp) {
		return false;
	}
```